### ISSUE TITLE
how to debug and modify

### ISSUE THREAD

**Issue by jly0810** [cgtuebingen/SIGNeRF#7]
Your work is built on nerfstudio. I am full of doubts about this because it does not have a main function entry like regular projects and can be easily debugged on pychar or VScode. I would like to ask you how to debug your work? I want to modify the virtual camera to off axis shooting mode during the inference phase (i.e. add an offset to the constructed light), how can I modify it in the code?

**Comment by JDihlmann** [comment]
Yes you are right, nerfstudio is great as a framework but hard to debug. So what I did during development was installing nerfstudio from source as follows, and log or change within the specific parts I need.

```
git clone https://github.com/nerfstudio-project/nerfstudio.git
cd nerfstudio
pip install --upgrade pip setuptools
pip install -e .
```

You can fully abstract the nerfstudio pipeline and replace the components with parts that add your functionality (e.g. we did this with the dataparser). The question is which inference mode are you referring to? If you are talking about the image / video rendering with ns-render you could simply specify the camera intrinsics and off axis shooting mode within the transforms.json you define as camera path. If you referring to inference the viewer, I'm afraid you need to look into the viewer code of nerfstudio to change the camera. Does this answer your question?

**Comment by jly0810** [comment]
Thank you for your reply! 1. I want to modify the direction of the light in the inference stage of the model to conform to the off-axis shooting method, so that the image generated by the horizontally arranged virtual camera array conforms to the off-axis shooting effect. Where should I make the modifications? In the original nerf code, I implemented it through the following code. I didn't understand the meaning of your modification in transforms.json. 2. How to integrate nerfstudio source code modification with your work signerf? How to make Signerf work with the modified nerfstudio?

rays_o, rays_d = get_rays_jj(test_dataset.directions, c2w, center)
def get_rays_jj(directions, c2w, center_c2w):
    """
    Get ray origin and normalized directions in world coordinate for all pixels in one image.

    Inputs:
        directions: (H, W, 3) precomputed ray directions in camera coordinate
        c2w: (3, 4) transformation matrix from camera coordinate to world coordinate
    Outputs:
        rays_o: (H*W, 3), the origin of the rays in world coordinate
        rays_d: (H*W, 3), the normalized direction of the rays in world coordinate
    """
    # Rotate ray directions from camera coordinate to the world coordinate
    rays_d = directions @ c2w[:3, :3].T  # (H, W, 3)
    # rays_d = rays_d / torch.norm(rays_d, dim=-1, keepdim=True)
    # The origin of all rays is the camera origin in world coordinate
    rays_o = c2w[:3, 3].expand(rays_d.shape)  # (H, W, 3)
    center_rays_o = center_c2w[:3, 3].expand(rays_d.shape)  # (H, W, 3)

    tt = center_rays_o - rays_o  # 当前相机指向中心相机间的偏移向量

    rays_d = rays_d + tt  # 当前相机的原始方向+偏移向量=当前相机模拟离轴变换之后的光线方向
    rays_d = rays_d.view(-1, 3)
    rays_o = rays_o.view(-1, 3)

    return rays_o, rays_d

**Comment by JDihlmann** [comment]
Okay I see. Here is what you could do for **inference**. Inference meaning you are running ns-render ... (which calls nerfstudio/scripts/render.py). Install nerfstudio as I outlined above in your current conda / python environment. You can then modify the nerfstudio repository  you just downloaded  and it will affect SIGNeRF as it is based on this nerfstudio version after your installation. In nerfstudio/scripts/render.py in line 154 you find the ray  bundle function either you change the entire class or just edit the current bundle: 


 ```
for i in range(len(train_cameras)):
      train_cam_pos = train_cameras[i].camera_to_worlds[:, 3].cpu()
      # Make sure the line of sight from rendered cam to training cam is not blocked by any object
      bundle = RayBundle(
          origins=cam_pos.view(1, 3),
          directions=((cam_pos - train_cam_pos) / (cam_pos - train_cam_pos).norm()).view(1, 3),
          pixel_area=torch.tensor(1).view(1, 1),
          nears=torch.tensor(0.05).view(1, 1),
          fars=torch.tensor(100).view(1, 1),
          camera_indices=torch.tensor(0).view(1, 1),
          metadata={},
      ).to(pipeline.device)
      outputs = pipeline.model.get_outputs(bundle)
      
      ...
```
